source/fundamentals/crud/write-operations/pkFactory.txt

```diff
@@ -0,0 +1,124 @@
+.. _node-pkfactory:
+
+=================================
+Generate Custom Vales for ``_id``
+=================================
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to use the MongoDB Node.js driver to generate your
+own ``_id`` values using the drivers ``pkFactory`` option.
+
+Behavior
+--------
+
+The ``pkFactory``, primary key factory option defines a system for
+creating unique identifiers in your documents. 
+
+The driver uses the primary key factory to generate unique ``_id``
+values for your documents when you choose not to specify an ``_id``
+during an :ref:`insert operation <node-fundamentals-insert-data>`. The
+default primary key factory generates ``ObjectId`` values.
+
+
+.. note:: Upsert Operations
+
+   The driver doesn't use any ``pkFactory`` methods for :ref:`upsert
+   operations <node-fundamentals-upsert>`. Instead, the driver always 
+   creates ``ObjectId`` values for the ``_id`` field of the
+   upserted documents.
+
+   If you want to use ``pkFactory`` methods, peform a :ref:`find
+   operation <node-fundamentals-retrieve-data>`, then an :ref:`update
+   <node-fundamentals-change-a-document>` or :ref:`insert
+   <node-fundamentals-insert-data>` operation.
+
+Data Consistency
+----------------
+
+If you do not specify a primary key factory method that generates
+``_id`` values of your desired type, you may experience data
+inconsistency.
+
+This may occur because the driver's default primary factory method generates
+``_id`` values of type ``ObjectId`` regardless of the type you specify for the 
+``_id`` field in your type parameter.
+
+Specify a ``pkFactory``
+-----------------------
+
+To keep the ``_id`` consistent, as a ``ObjectID``, apply a primary key
+factory to your ``MongoClient`` instance like this:
+
+.. code-block:: javascript
+
+   const uuid = require('bson');
+   ...
+   const client = new MongoClient(uri, { pkFactory: { createPk: () => uuid }});
+
+Example
+~~~~~~~
+
+Assume you have a collection, ``characters.villans``, that contains
+documents with a villans name. The following example inserts
+documents with an ``_id`` of type ``ObjectID`` into your collection:
+
+.. code-block:: javascript
+
+   const uuid = require('bson');
+   const { MongoClient } = require("mongodb");
+
```

Comment:
issue: `villain` is misspelled.
issue: Will it really insert documents with type `ObjectID`? Isn't that the default behavior? We should show users non-default behavior.
issue: Does the example need to be so long? It seems we can be matter-of-fact about it in the previous section and not show a full example.

Reply:
Apologies, I didn't realize the `_id` would be saved as an `ObjectID` regardless of the type we specify in the `pkFactory` option. And yes, I agree that the example should be removed as it's only implementing the code snippet right above it.